halo2-base/src/gates/builder/mod.rs

```diff
@@ -586,41 +594,48 @@ pub struct RangeCircuitBuilder<F: ScalarField>(pub GateCircuitBuilder<F>);
 
 impl<F: ScalarField> RangeCircuitBuilder<F> {
     /// Creates an instance of the [RangeCircuitBuilder] and executes in keygen mode.
-    pub fn keygen(builder: GateThreadBuilder<F>) -> Self {
-        Self(GateCircuitBuilder::keygen(builder))
+    pub fn keygen(builder: GateThreadBuilder<F>, config_params: BaseConfigParams) -> Self {
+        Self(GateCircuitBuilder::keygen(builder, config_params))
     }
 
     /// Creates a mock instance of the [RangeCircuitBuilder].
-    pub fn mock(builder: GateThreadBuilder<F>) -> Self {
-        Self(GateCircuitBuilder::mock(builder))
+    pub fn mock(builder: GateThreadBuilder<F>, config_params: BaseConfigParams) -> Self {
+        Self(GateCircuitBuilder::mock(builder, config_params))
     }
 
     /// Creates an instance of the [RangeCircuitBuilder] and executes in prover mode.
     pub fn prover(
         builder: GateThreadBuilder<F>,
+        config_params: BaseConfigParams,
         break_points: MultiPhaseThreadBreakPoints,
     ) -> Self {
-        Self(GateCircuitBuilder::prover(builder, break_points))
+        Self(GateCircuitBuilder::prover(builder, config_params, break_points))
     }
 }
 
 impl<F: ScalarField> Circuit<F> for RangeCircuitBuilder<F> {
     type Config = BaseConfig<F>;
     type FloorPlanner = SimpleFloorPlanner;
+    type Params = BaseConfigParams;
+
+    fn params(&self) -> Self::Params {
+        self.0.config_params.clone()
+    }
 
     /// Creates a new instance of the [RangeCircuitBuilder] without witnesses by setting the witness_gen_only flag to false
     fn without_witnesses(&self) -> Self {
         unimplemented!()
     }
 
     /// Configures a new circuit using [`BaseConfigParams`]
-    fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
-        let params = BASE_CONFIG_PARAMS
-            .try_with(|config| config.borrow().clone())
-            .expect("You need to call config() to configure the halo2-base circuit shape first");
+    fn configure_with_params(meta: &mut ConstraintSystem<F>, params: Self::Params) -> Self::Config {
         BaseConfig::configure(meta, params)
     }
 
+    fn configure(_: &mut ConstraintSystem<F>) -> Self::Config {
```

Comment:
In new `halo2_proofs`, when features "circuit-params" is on (which we always turn on in halo2-lib), `Circuit` will call `configure_with_params` instead of `configure`. I can just remove `configure` in the trait when the feature flag is on, but since this wasn't done in upstream I left it.

Reply:
I changed to `unreachable` to make this clearer